# Euro-millions Number analysis
Read out the historic data to base our new numbers with and perform some data exploration.

In [1]:
import pandas as pd

df = pd.read_excel("lottery-numbers-tracker.xlsx", sheet_name="EuroMillions", header=0)

df.head()


DrawDate  Ball 1  Ball 2  Ball 3  Ball 4  Ball 5  Lucky Star 1  \
0  2023-10-03 00:00:00       6      20      22      24      45             4   
1  2023-09-29 00:00:00       9      11      13      21      32             2   
2  2023-09-26 00:00:00       2       6      14      19      23             5   
3  2023-09-22 00:00:00       3      23      24      34      35             5   
4  2023-09-19 00:00:00      10      15      31      41      42             2   

   Lucky Star 2  
0             5  
1             7  
2             7  
3             8  
4             5

The probability for each ball to be drawn given that there is no bias is:

In [2]:
probabilities = {i+1: 1/50 for i in range(50)}
print(probabilities)
probabilities_ls = {i+1: 1/12 for i in range(12)}
print(probabilities_ls)

{1: 0.02, 2: 0.02, 3: 0.02, 4: 0.02, 5: 0.02, 6: 0.02, 7: 0.02, 8: 0.02, 9: 0.02, 10: 0.02, 11: 0.02, 12: 0.02, 13: 0.02, 14: 0.02, 15: 0.02, 16: 0.02, 17: 0.02, 18: 0.02, 19: 0.02, 20: 0.02, 21: 0.02, 22: 0.02, 23: 0.02, 24: 0.02, 25: 0.02, 26: 0.02, 27: 0.02, 28: 0.02, 29: 0.02, 30: 0.02, 31: 0.02, 32: 0.02, 33: 0.02, 34: 0.02, 35: 0.02, 36: 0.02, 37: 0.02, 38: 0.02, 39: 0.02, 40: 0.02, 41: 0.02, 42: 0.02, 43: 0.02, 44: 0.02, 45: 0.02, 46: 0.02, 47: 0.02, 48: 0.02, 49: 0.02, 50: 0.02}
{1: 0.08333333333333333, 2: 0.08333333333333333, 3: 0.08333333333333333, 4: 0.08333333333333333, 5: 0.08333333333333333, 6: 0.08333333333333333, 7: 0.08333333333333333, 8: 0.08333333333333333, 9: 0.08333333333333333, 10: 0.08333333333333333, 11: 0.08333333333333333, 12: 0.08333333333333333}


Now we sum the number of occurrences for each ball using numpy's unique function:

In [3]:
import numpy as np
n = df.shape[0] * 5
occurrences = np.concatenate([df["Ball 1"], df["Ball 2"], df["Ball 3"], df["Ball 4"], df["Ball 5"]])

unique, counts = np.unique(occurrences, return_counts=True)
print(unique, len(unique))
running_counts = dict(zip(unique, counts))
print(running_counts)


[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50] 50
{1: 10, 2: 14, 3: 19, 4: 9, 5: 14, 6: 19, 7: 16, 8: 14, 9: 16, 10: 20, 11: 19, 12: 18, 13: 20, 14: 16, 15: 14, 16: 17, 17: 21, 18: 15, 19: 20, 20: 15, 21: 26, 22: 10, 23: 22, 24: 20, 25: 22, 26: 17, 27: 19, 28: 16, 29: 18, 30: 9, 31: 18, 32: 18, 33: 17, 34: 27, 35: 27, 36: 17, 37: 17, 38: 16, 39: 11, 40: 16, 41: 12, 42: 19, 43: 15, 44: 20, 45: 20, 46: 12, 47: 18, 48: 24, 49: 12, 50: 14}


Same unique counts for the lucky star balls:

In [4]:
n_ls = df.shape[0] * 2
occurrences_ls = np.concatenate([df["Lucky Star 1"], df["Lucky Star 2"]])

unique_ls, counts_ls = np.unique(occurrences_ls, return_counts=True)
print(unique_ls, len(unique_ls))
running_counts_ls = dict(zip(unique_ls, counts_ls))
print(running_counts_ls)


[ 1  2  3  4  5  6  7  8  9 10 11 12] 12
{1: 21, 2: 31, 3: 39, 4: 19, 5: 26, 6: 31, 7: 28, 8: 26, 9: 22, 10: 35, 11: 40, 12: 24}


Now we calculate the probability of each number and how it deviates from the raw unbiased probabilities calculated previously. It turns out the standard deviation is around a quarter of the mean, and by using the central limit theorem, we get the s.d. of the true distribution as 

In [5]:
import math
running_probabilities = {i+1: (running_counts[i + 1] / n) for i in range(50)}
print(running_probabilities)
ave = sum([running_probabilities[i + 1] for i in range(50)]) / 50.0
sd = math.sqrt(sum([(ave - running_probabilities[i + 1])**2 for i in range(50)]) / 50.0)
print(ave, sd)
print("s.d. of the true distribution", sd/math.sqrt(len(df[0:])))

{1: 0.011695906432748537, 2: 0.016374269005847954, 3: 0.022222222222222223, 4: 0.010526315789473684, 5: 0.016374269005847954, 6: 0.022222222222222223, 7: 0.01871345029239766, 8: 0.016374269005847954, 9: 0.01871345029239766, 10: 0.023391812865497075, 11: 0.022222222222222223, 12: 0.021052631578947368, 13: 0.023391812865497075, 14: 0.01871345029239766, 15: 0.016374269005847954, 16: 0.019883040935672516, 17: 0.02456140350877193, 18: 0.017543859649122806, 19: 0.023391812865497075, 20: 0.017543859649122806, 21: 0.0304093567251462, 22: 0.011695906432748537, 23: 0.025730994152046785, 24: 0.023391812865497075, 25: 0.025730994152046785, 26: 0.019883040935672516, 27: 0.022222222222222223, 28: 0.01871345029239766, 29: 0.021052631578947368, 30: 0.010526315789473684, 31: 0.021052631578947368, 32: 0.021052631578947368, 33: 0.019883040935672516, 34: 0.031578947368421054, 35: 0.031578947368421054, 36: 0.019883040935672516, 37: 0.019883040935672516, 38: 0.01871345029239766, 39: 0.012865497076023392, 40

This suggests there is an inherent bias in the sample of around 4 in 10000, and is consistent with why the probability of a win is a few order of magnitudes larger than it is supposed to be from a purely theoretical evaluation.
Same calculation for the lucky star balls suggests a much larger bias here:

In [6]:
running_probabilities_ls = {i+1: (running_counts_ls[i + 1] / n_ls) for i in range(12)}
print(running_probabilities_ls)
ave_ls = sum([running_probabilities_ls[i + 1] for i in range(12)]) / 12.0
sd_ls = math.sqrt(sum([(ave_ls - running_probabilities_ls[i + 1])**2 for i in range(12)]) / 12.0)
print(ave_ls, sd_ls)
print("s.d. of the true distribution", sd_ls/math.sqrt(len(df[0:])))


{1: 0.06140350877192982, 2: 0.09064327485380116, 3: 0.11403508771929824, 4: 0.05555555555555555, 5: 0.07602339181286549, 6: 0.09064327485380116, 7: 0.08187134502923976, 8: 0.07602339181286549, 9: 0.06432748538011696, 10: 0.1023391812865497, 11: 0.11695906432748537, 12: 0.07017543859649122}
0.08333333333333333 0.0192294538566753
s.d. of the true distribution 0.0014705130894864227


Now we evaluate how far apart each ball is compared to the truly unbiased mid-points of a perfectly unbiased distribution. ave_dv below is what the separation is supposed to be in a non-biased distribution but sd_dv is how that figure deviates from the ideal case, square rooted to accommodate the deviation in either direction. And it turns out the numbers tend to be less spread out from each other than the non-biased case, by around a quarter.

In [7]:
dv = 0
ddv = 0
for i in range(int(n / 5)):
    dv += sum([
        df["Ball 1"][i],
        df["Ball 2"][i] - df["Ball 1"][i],
        df["Ball 3"][i] - df["Ball 2"][i],
        df["Ball 4"][i] - df["Ball 3"][i],
        df["Ball 5"][i] - df["Ball 4"][i],
        50 - df["Ball 5"][i]
    ])
ave_dv = dv / (n * 6 / 5)

for i in range(int(n / 5)):
    ddv += sum([
        (ave_dv - df["Ball 1"][i])**2,
        (ave_dv - (df["Ball 2"][i] - df["Ball 1"][i]))**2,
        (ave_dv - (df["Ball 3"][i] - df["Ball 2"][i]))**2,
        (ave_dv - (df["Ball 4"][i] - df["Ball 3"][i]))**2,
        (ave_dv - (df["Ball 5"][i] - df["Ball 4"][i]))**2,
        (ave_dv - (50 - df["Ball 5"][i]))**2
    ])
sd_dv = math.sqrt(ddv / (n * 6 / 5))
print(ave_dv, sd_dv)


8.333333333333334 6.408753760001652


Repeat the same for lucky star balls, and it turns out there is even more bias, around a third: 

In [8]:
dv_ls = 0
ddv_ls = 0
for i in range(int(n_ls / 2)):
    dv_ls += sum([
        df["Lucky Star 1"][i],
        df["Lucky Star 2"][i] - df["Lucky Star 1"][i],
        12 - df["Lucky Star 2"][i]
    ])
ave_dv_ls = dv_ls / (n_ls * 3 / 2)

for i in range(int(n_ls / 2)):
    ddv_ls += sum([
        (ave_dv_ls - df["Lucky Star 1"][i])**2,
        (ave_dv_ls - (df["Lucky Star 2"][i] - df["Lucky Star 1"][i]))**2,
        (ave_dv_ls - (12 - df["Lucky Star 2"][i]))**2
    ])
sd_dv_ls = math.sqrt(ddv_ls / (n_ls * 3 / 2))
print(ave_dv_ls, sd_dv_ls)

4.0 2.7137653577923784


One way to make sense of this is to consider the case for lucky star balls. In a situation where there is no bias and there are infinite samples to choose from, the average probability of getting the two balls will be 4 and 8, 4 representing the smaller ball values and 8 the bigger ones. In this case the average outcome of the balls are perfectly and evenly skewed, and the distance between each draw and its neighbour is 4. that is to say 0-4, 4-8, 8-12 all have neighbouring distances of 4. 

Now consider our case where this figure is closer to 2.71, this means it is more likely to get a neighbouring average figure of 2.71, so it is more likely to get numbers like 0-2, 2-4, 4-12, which would give `sqrt((4-2)**2 + (4-2)**2 + (4-8)**2 / 3) ~ 2.82`

Now we look at repetitions in the past 10 draws for each ball:

In [9]:
rep_counts = {i+1: 0 for i in range(50)}
total_reps = 0
for i in range(int(n / 5) - 10):
    for j in range(10):
        for k in range(5):
            if (any([
                df["Ball {}".format(k + 1)][i] == df["Ball 1"][i + j + 1],
                df["Ball {}".format(k + 1)][i] == df["Ball 2"][i + j + 1],
                df["Ball {}".format(k + 1)][i] == df["Ball 3"][i + j + 1],
                df["Ball {}".format(k + 1)][i] == df["Ball 4"][i + j + 1],
                df["Ball {}".format(k + 1)][i] == df["Ball 5"][i + j + 1]
            ])):
                rep_counts[df["Ball {}".format(k + 1)][i]] += 1
                total_reps += 1
print(rep_counts)
rep_averages = {}
for i in range(50):
    rep_averages[i + 1] = rep_counts[i + 1] / total_reps
print(rep_averages)
rep_ave = sum([rep_averages[i + 1] for i in range(50)]) / 50.0
rep_sd = math.sqrt(sum([(rep_ave - rep_averages[i + 1])**2 for i in range(50)]) / 50.0)
print(rep_ave, rep_sd)

{1: 7, 2: 8, 3: 14, 4: 4, 5: 10, 6: 12, 7: 12, 8: 13, 9: 10, 10: 27, 11: 23, 12: 12, 13: 16, 14: 11, 15: 7, 16: 22, 17: 24, 18: 17, 19: 22, 20: 13, 21: 42, 22: 5, 23: 24, 24: 23, 25: 20, 26: 13, 27: 19, 28: 22, 29: 13, 30: 4, 31: 15, 32: 16, 33: 14, 34: 36, 35: 51, 36: 14, 37: 15, 38: 8, 39: 6, 40: 15, 41: 12, 42: 18, 43: 14, 44: 22, 45: 17, 46: 9, 47: 16, 48: 36, 49: 6, 50: 11}
{1: 0.00853658536585366, 2: 0.00975609756097561, 3: 0.01707317073170732, 4: 0.004878048780487805, 5: 0.012195121951219513, 6: 0.014634146341463415, 7: 0.014634146341463415, 8: 0.015853658536585366, 9: 0.012195121951219513, 10: 0.032926829268292684, 11: 0.02804878048780488, 12: 0.014634146341463415, 13: 0.01951219512195122, 14: 0.013414634146341463, 15: 0.00853658536585366, 16: 0.026829268292682926, 17: 0.02926829268292683, 18: 0.020731707317073172, 19: 0.026829268292682926, 20: 0.015853658536585366, 21: 0.05121951219512195, 22: 0.006097560975609756, 23: 0.02926829268292683, 24: 0.02804878048780488, 25: 0.024390

Now the probability of a repetition for each ball is 0.2 * 1/10 attempts so roughly 0.02 which is very close to the average evaluated. However, the standard deviation is very large here, showing a tendency of numbers to repeat themselves in these smaller cycles. 

Repeating the same operation for lucky star numbers:

In [10]:
rep_counts_ls = {i+1: 0 for i in range(12)}
total_reps_ls = 0
for i in range(int(n / 5) - 10):
    for j in range(10):
        for k in range(2):
            if (any([
                df["Lucky Star {}".format(k + 1)][i] == df["Lucky Star 1"][i + j + 1],
                df["Lucky Star {}".format(k + 1)][i] == df["Lucky Star 2"][i + j + 1],
            ])):
                rep_counts_ls[df["Lucky Star {}".format(k + 1)][i]] += 1
                total_reps_ls += 1
print(rep_counts_ls)
rep_averages_ls = {}
for i in range(12):
    rep_averages_ls[i + 1] = rep_counts_ls[i + 1] / total_reps_ls
print(rep_averages_ls)
rep_ave_ls = sum([rep_averages_ls[i + 1] for i in range(12)]) / 12.0
rep_sd_ls = math.sqrt(sum([(rep_ave_ls - rep_averages_ls[i + 1])**2 for i in range(12)]) / 12.0)
print(rep_ave_ls, rep_sd_ls)


{1: 36, 2: 43, 3: 99, 4: 15, 5: 37, 6: 54, 7: 38, 8: 29, 9: 23, 10: 67, 11: 83, 12: 23}
{1: 0.06581352833638025, 2: 0.07861060329067641, 3: 0.1809872029250457, 4: 0.027422303473491772, 5: 0.06764168190127971, 6: 0.09872029250457039, 7: 0.06946983546617916, 8: 0.05301645338208409, 9: 0.04204753199268738, 10: 0.12248628884826325, 11: 0.15173674588665448, 12: 0.04204753199268738}
0.08333333333333333 0.04491369597338533


We see that the s.d. here is also quite large it is generally apparent that balls repeat themselves in cycles of 10. The degree to which this is true is still unknown but is possible to evaluate it if a running average of this figure is kept and its variance evaluated over a longer period.

Now we put together our findings and try to assign a weight to each of the evaluated probabilities in a way that the new guesses will get scores based on them. 

It is difficult to guess how much each probability would influence future events but this should be looked at and adjusted over time to monitor how the bias is shifting in the outcome distributions. Perhaps a GAN can be optimised to consider all of these factors in guessing of new numbers.

In [11]:
from random import randint
candidates = {}
for i in range(0, 1000):
    new_guess = []
    while len(new_guess) < 5:
        new_random = randint(1, 50)
        if new_random not in new_guess:
            new_guess.append(new_random)
    if (str(new_guess) in candidates):
        continue

    new_guess.sort()
    # print(new_guess)
    score = 0
    ave_score = 0
    rep_score = 0
    for j in range(len(new_guess)):
        if abs(running_probabilities[new_guess[j]] - ave) < (1.75*sd):
            ave_score += abs((abs(running_probabilities[new_guess[j]] - ave) - sd))
        else:
            ave_score += abs(running_probabilities[new_guess[j]] - ave)

        if (rep_averages[new_guess[j]] - rep_ave) < (1.45*rep_sd):
            rep_score += abs((abs(rep_averages[new_guess[j]] - rep_ave) - rep_sd))
        else:
            rep_score += abs(rep_averages[new_guess[j]] - rep_ave)
    # print(ave_score)
    # print(rep_score)

    guess_ave_dv = sum([
        new_guess[0],
        new_guess[1] - new_guess[0],
        new_guess[2] - new_guess[1],
        new_guess[3] - new_guess[2],
        new_guess[4] - new_guess[3],
        50 - new_guess[4]
    ]) / 6

    guess_ddv = sum([
        (guess_ave_dv - new_guess[0])**2,
        (guess_ave_dv - (new_guess[1] - new_guess[0]))**2,
        (guess_ave_dv - (new_guess[2] - new_guess[1]))**2,
        (guess_ave_dv - (new_guess[3] - new_guess[2]))**2,
        (guess_ave_dv - (new_guess[4] - new_guess[3]))**2,
        (guess_ave_dv - (50 - new_guess[4]))**2
    ])
    guess_sd_dv = math.sqrt(guess_ddv / 6)
    ddv_score = abs(sd_dv - guess_sd_dv) / (n / 5)
    # print(ddv_score)

    candidates[str(new_guess)] = ave_score + rep_score + ddv_score




In [12]:
candidates_ls = {}
for i in range(0, 1000):
    new_guess_ls = []
    while len(new_guess_ls) < 2:
        new_random = randint(1, 12)
        if new_random not in new_guess_ls:
            new_guess_ls.append(new_random)
    if (str(new_guess_ls) in candidates_ls):
        continue

    new_guess_ls.sort()
    # print(new_guess_ls)
    score = 0
    ave_score = 0
    rep_score = 0
    for j in range(len(new_guess_ls)):
        if abs(running_probabilities_ls[new_guess_ls[j]] - ave_ls) < (sd_ls):
            ave_score += abs((abs(running_probabilities_ls[new_guess_ls[j]] - ave_ls) - sd_ls))
        else:
            ave_score += abs(running_probabilities_ls[new_guess_ls[j]] - ave_ls)

        if (rep_averages_ls[new_guess_ls[j]] - rep_ave_ls) < (2*rep_sd_ls):
            rep_score += abs((abs(rep_averages_ls[new_guess_ls[j]] - rep_ave_ls) - rep_sd_ls))
        else:
            rep_score += abs(rep_averages_ls[new_guess_ls[j]] - rep_ave_ls)
    # print(ave_score)
    # print(rep_score)

    guess_ave_dv = sum([
        new_guess_ls[0],
        new_guess_ls[1] - new_guess_ls[0],
        12 - new_guess_ls[1]
    ]) / 3

    guess_ddv = sum([
        (guess_ave_dv - new_guess_ls[0])**2,
        (guess_ave_dv - (new_guess_ls[1] - new_guess_ls[0]))**2,
        (guess_ave_dv - (12 - new_guess_ls[1]))**2
    ])
    guess_sd_dv = math.sqrt(guess_ddv / 3)
    ddv_score = abs(sd_dv_ls - guess_sd_dv) / (n_ls / 2)
    # print(ddv_score)

    candidates_ls[str(new_guess_ls)] = ave_score + rep_score + ddv_score



200th to 240th best guesses for the ball numbers:

In [13]:
for i in sorted(candidates.items(), key=lambda item: item[1])[200:240]:
    print(i)

('[9, 13, 29, 49, 50]', 0.04564721331862864)
('[1, 24, 27, 30, 41]', 0.04568386981862649)
('[3, 20, 26, 39, 50]', 0.04570276199144559)
('[8, 12, 26, 28, 46]', 0.0457058052934905)
('[11, 30, 38, 45, 50]', 0.045795399038434575)
('[3, 9, 17, 20, 25]', 0.04583971369103265)
('[18, 22, 29, 45, 46]', 0.04594747044628069)
('[4, 12, 22, 24, 29]', 0.04598175137491835)
('[5, 9, 22, 38, 45]', 0.04600479141056466)
('[5, 9, 17, 28, 45]', 0.046056372453571935)
('[23, 26, 28, 43, 49]', 0.04608366356684283)
('[1, 2, 12, 37, 39]', 0.04614364053494007)
('[2, 19, 24, 25, 47]', 0.04616014068899912)
('[20, 24, 26, 43, 46]', 0.046202414692968194)
('[1, 7, 19, 32, 46]', 0.04621793230876897)
('[3, 10, 16, 22, 38]', 0.04629720535201941)
('[14, 22, 36, 44, 49]', 0.04638141231181668)
('[7, 9, 12, 33, 44]', 0.046492182404691475)
('[6, 7, 22, 25, 49]', 0.046613662754910676)
('[6, 8, 22, 28, 31]', 0.04662070269822185)
('[8, 26, 41, 42, 44]', 0.04676783944129877)
('[23, 26, 28, 39, 47]', 0.04677244827258653)
('[2, 6,

top guesses for the lucky star numbers:

In [14]:
for i in sorted(candidates_ls.items(), key=lambda item: item[1]):
    print(i)

('[9, 10]', 0.014778892080674389)
('[9, 12]', 0.019562018785739092)
('[10, 12]', 0.02507987286972782)
('[8, 9]', 0.03171376416345163)
('[8, 10]', 0.03317121136911059)
('[8, 12]', 0.03944514448470868)
('[4, 10]', 0.05107976326554557)
('[4, 12]', 0.05170392779059914)
('[6, 12]', 0.05181623953978101)
('[5, 12]', 0.05218694589902152)
('[6, 9]', 0.052897467776421195)
('[4, 9]', 0.053721751246424435)
('[6, 10]', 0.0537509007396673)
('[1, 9]', 0.05452115134294384)
('[5, 10]', 0.05472562134132066)
('[7, 9]', 0.055906114243684246)
('[5, 9]', 0.0560881964597296)
('[7, 10]', 0.05803896006940027)
('[2, 10]', 0.05876536127770291)
('[2, 9]', 0.059198928851807234)
('[7, 12]', 0.05986305268029524)
('[1, 10]', 0.060251107082311836)
('[9, 11]', 0.06590999329284793)
('[2, 12]', 0.07120600477474937)
('[10, 11]', 0.07204058533557475)
('[1, 12]', 0.07219737660497794)
('[6, 8]', 0.07428288273609651)
('[7, 8]', 0.07687946344995879)
('[1, 8]', 0.07738551532898166)
('[5, 8]', 0.07875302428187485)
('[11, 12]', 0

# GAN timeseries generation

In this section we will extract the data to be loaded into a neural net to make predictions and compare them to the existing ones we evaluated here. Eventually the idea is to use a GAN to enhance this prediction.

In [15]:
df[["Lucky Star 1", "Lucky Star 2", "Ball 1", "Ball 2", "Ball 3", "Ball 4", "Ball 5"]].to_csv("euromillions-dataset.csv", index=None)